src/respec.js

```diff
@@ -0,0 +1,25 @@
+import { runAll } from "./core/base-runner.js";
+import { ui } from "./core/ui.js";
+
+// In case everything else fails, we want the error
+window.addEventListener("error", ev => {
+  console.error(ev.error, ev.message, ev);
+});
+
+export async function run(plugins) {
+  try {
+    ui.show();
+    await domReady();
+    await runAll(plugins);
+  } finally {
+    ui.enable();
+  }
+}
+
```

Comment:
I'm not a huge fan of mixing together the running the plugins and showing the UI together. I think they should be separate concerns. Also, we should only allow plugins to be run once. 

Could we do something like: 

```suggestion
import { runAll } from "./core/base-runner.js";
import { ui } from "./core/ui.js";

// In case everything else fails, we want the error
window.addEventListener("error", ev => {
  console.error(ev.error, ev.message, ev);
});

let initialized = false;
let pluginsDone; 
const pluginsPromise = new Promise((res) => {
   pluginsDone = res;
});
domReady().then(async () => {
   ui.show();
   await pluginsPromise;
   ui.enable();
});

export async function run(plugins) {
  if (initialized) throw new Error("ReSpec's was already initialized!");
  await domReady();
  await runAll(plugins);
  initialized = true;
  pluginsDone();
}
```

Reply:
Any thoughts on the above @sidvishnoi?

Reply:
> Also, we should only allow plugins to be run once.

The way initialization code runs, it's unlikely that plugins can run multiple times. I'm not sure what the concern is here.

> I'm not a huge fan of mixing together the running the plugins and showing the UI together.

I think the mixing is not bad here. The tasks are related and have a sound flow. Unless we add more things like ui, I think the current approach is fine.

Reply:
>  I'm not sure what the concern is here.

I'm worried that because this is the most "public" part of the API, random documents/folks will try to run their own plugins through it more than once. We've seen the "creative" things people do in their own documents when interacting with ReSpec, so I'd like to ensure that never happens.

Reply:
IMO, if they're calling `respec.run(plugins)` multiple times in their profile, nothing stops them from writing their own runners. The fault is on their end and we shouldn't mess with our API to avoid that.

Though, if we're to provide a safer API, we should _probably_ wrap it in a class:
```js
// file: respec.js
class ReSpec {
  constructor(plugins) {
    this.#initialized = false;
  }

  async run() {
    // ...
  }

  // maybe more methods like getPluginRunDurations()?
}

// file: [profile].js
import ReSpec from "./respec.js";
const respec = new ReSpec(plugins);
await respec.run();
```

Reply:
Though, if we're to provide a safer API, we should _probably_ wrap it in a class:
```js
// file: respec.js
class ReSpec {
  constructor(plugins) {
    this.#initialized = false;
  }

  async run() {
    // ...
  }

  // maybe more methods like getPluginRunDurations()?
}

// file: [profile].js
import ReSpec from "./respec.js";
const respec = new ReSpec(plugins);
await respec.run();
```

Edit: class isn't that useful, as a module with global state is similar.

Reply:
> IMO, if they're calling respec.run(plugins) multiple times in their profile, nothing stops them from writing their own runners. 

It's not in the profile creators I'm worried about (as we review those, so we are protected). I'm worried about random spec editors fining it and using it. Get what I mean?

What about the separation of concerns about UI and plugins part? Can we separate them or are they inherently tied together?

Reply:
The `respec.run` isn't exposed to webpage. It's part of the bundling process. So, that should be fine?

> What about the separation of concerns about UI and plugins part? Can we separate them or are they inherently tied together?

They look tied together to me, as in: show loader -> run plugins -> show UI.